### PR TITLE
[GPU][QWen2-VL][QWen2.5-VL] improve SDPA performance with cu_seqlens and cu_window_seqlens

### PR DESCRIPTION
**Pull request by ceciliapeng2011**
### Details:
 The process is like,

1.  GenAI provides a special RT_INFO entry to QWen-VL models during compile_model.
2.  The plugin detects this entry and the target device capability.
3.  The plugin transforms the model input, replacing attention_mask with cu_seqlens. 
4.  GenAI then performs inference after validating the final model inputs.

### Tickets:
 - *[168519](https://jira.devtools.intel.com/browse/CVS-168519)*

Should work along with https://github.com/openvinotoolkit/openvino/pull/30909

### CODE REVIEW

File: src/cpp/src/visual_language/qwen2_5_vl/classes.cpp [hunk 1]
```diff
@@ -152,13 +163,20 @@ ov::Tensor InputsEmbedderQwen2_5_VL::run_image_embeddings_merger(
...
     ov::InferRequest& vision_embeddings_merger = infer_request_guard.get();
     vision_embeddings_merger.set_tensor("hidden_states", concatenated_embeds);
-    vision_embeddings_merger.set_tensor("attention_mask", attention_mask);
+    if (m_with_cu_seqlens_input) {
+        vision_embeddings_merger.set_tensor("cu_seq_lens", cu_seq_lens);
+        vision_embeddings_merger.set_tensor("cu_window_seqlens", t_cu_window_seqlens);
+    }
+    else {
+        vision_embeddings_merger.set_tensor("attention_mask", attention_mask);
+        vision_embeddings_merger.set_tensor("window_attention_mask", window_attention_mask);
```
Comment (Wovchena) [inline src/cpp/src/visual_language/qwen2_5_vl/classes.cpp:177]: Move computation of corresponding Tensors under `if` or `else` to avoid computing both of the paths

File: src/cpp/src/visual_language/qwen2vl/classes.cpp [hunk 2]
```diff
@@ -319,6 +320,29 @@ ov::Tensor get_attention_mask(const std::vector<std::array<size_t, 3>>& reordere
...
+    }
+
+    ov::Tensor t_cu_seqlens = ov::Tensor(ov::element::i32, {cu_seqlens.size()});
+    auto* ptr = static_cast<int32_t*>(t_cu_seqlens.data());
+    for (size_t n = 0; n < cu_seqlens.size(); n++) {
+        ptr[n] = cu_seqlens[n];
+    }
+    return t_cu_seqlens;
+
+    // return ov::Tensor(ov::element::i32, {cu_seqlens.size()}, cu_seqlens.data());
```
Comment (Wovchena) [inline src/cpp/src/visual_language/qwen2vl/classes.cpp]:
```suggestion
```

File: src/cpp/src/visual_language/vl_sdpa_transformations.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "visual_language/vl_sdpa_transformations.hpp"
+
+namespace ov {
+namespace genai {
+namespace utils {
+
+void apply_vl_sdpa_transformations(std::shared_ptr<ov::Model> model) {
+    const ov::op::util::VariableVector& variables = model->get_variables();
+    OPENVINO_ASSERT(variables.empty(), "Model is supposed to be stateless");
```
Comment (Wovchena) [inline src/cpp/src/visual_language/vl_sdpa_transformations.cpp]: I'd propose to move this check to the transformation because it's this transformation prerequisite, apparently

File: src/cpp/src/visual_language/qwen2vl/classes.cpp [hunk 4]
```diff
@@ -492,8 +516,16 @@ InputsEmbedderQwen2VL::InputsEmbedderQwen2VL(
...
-    auto compiled_model = utils::singleton_core().compile_model(model_dir / "openvino_vision_embeddings_merger_model.xml", device, device_config);
+    auto model = utils::singleton_core().read_model(model_dir / "openvino_vision_embeddings_merger_model.xml");
+    utils::apply_vl_sdpa_transformations(model);
+
+    auto compiled_model = utils::singleton_core().compile_model(model, device, device_config);
     ov::genai::utils::print_compiled_model_properties(compiled_model, "VLM vision embeddings merger model");
+
+    if (utils::check_vl_sdpa_transformations(compiled_model)) {
+        m_with_cu_seqlens_input = true;
+    }
```
Comment (Wovchena) [inline src/cpp/src/visual_language/qwen2vl/classes.cpp]:
```suggestion
    m_with_cu_seqlens_input = utils::check_vl_sdpa_transformations(compiled_model));
```
Same below

File: src/cpp/src/visual_language/vl_sdpa_transformations.hpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace ov {
+namespace genai {
+
+namespace utils {
+
+/** Applies transformations to the ov::Model to transform SDPA to VLSDPA.
+ * @param model Pointer to the ov::Model representing one of the supported VLM architectures.
+ */
+void apply_vl_sdpa_transformations(std::shared_ptr<ov::Model> model);
```
Comment (Wovchena) [inline src/cpp/src/visual_language/vl_sdpa_transformations.hpp]:
```suggestion
/** Requests transforming SDPA ov::Model to VLSDPA. It's up to a plugin to apply the transformation.
 * @param model Pointer to the ov::Model representing one of the supported VLM architectures.
 */
void request_vl_sdpa_transformations(std::shared_ptr<ov::Model> model);
```

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/visual_language/qwen2vl/classes.cpp [hunk 6]
```diff
@@ -627,12 +665,22 @@ ov::Tensor InputsEmbedderQwen2VL::run_image_embeddings_merger(
...
     ov::Tensor attention_mask = qwen2_vl_utils::get_attention_mask(reordered_images_grid_thw);
+    ov::Tensor cu_seq_lens = qwen2_vl_utils::get_cu_seqlens(reordered_images_grid_thw);
     ov::Tensor rotary_pos_emb = get_rotary_pos_emb(reordered_images_grid_thw);
 
+    // {
+    //     int32_t* ptr = static_cast<int32_t*>(cu_seq_lens.data());
+    //     for (auto i = 0; i < cu_seq_lens.get_size(); i++)
+    //         std::cout << "=========== cu_seq_lens i: " << i << " : " << ptr[i] << std::endl; 
+    // }
+
```
Comment (rkazants) [inline src/cpp/src/visual_language/qwen2vl/classes.cpp]:
```suggestion
```

File: src/cpp/src/visual_language/qwen2vl/classes.cpp [hunk 7]
```diff
@@ -319,6 +320,29 @@ ov::Tensor get_attention_mask(const std::vector<std::array<size_t, 3>>& reordere
...
+            cumsum += slice_len;
+            cu_seqlens.push_back(cumsum);
+        }
+    }
+
+    ov::Tensor t_cu_seqlens = ov::Tensor(ov::element::i32, {cu_seqlens.size()});
+    auto* ptr = static_cast<int32_t*>(t_cu_seqlens.data());
+    for (size_t n = 0; n < cu_seqlens.size(); n++) {
+        ptr[n] = cu_seqlens[n];
+    }
```
Comment (rkazants) [inline src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: I think we can create OV sub-graph for computation of `cu_seqlens` from `reordered_images_grid_thw` and attach to the main graph. I don't want us to have more computation in C++, rather to offload this work to plugin side.

File: src/cpp/src/visual_language/vl_sdpa_transformations.hpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+bool check_vl_sdpa_transformations(ov::CompiledModel& compiled_model);
```
Comment (rkazants) [inline src/cpp/src/visual_language/vl_sdpa_transformations.hpp]:
```suggestion
bool check_vl_sdpa_transformations(const ov::CompiledModel& compiled_model);
```

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 7 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: hmm, sounds good. However, to convert attention_mask to cu_seqlens, we still have to compute attention_mask in C++ code, and cost more memory.

Comment (rkazants) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: Not sure in memory increase. If you have it, then you have it in your C++ implementation too.
Please elaborate concern.

File: src/cpp/src/visual_language/qwen2_5_vl/classes.cpp (hunk 1 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2_5_vl/classes.cpp:177]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 2 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp]: done

File: src/cpp/src/visual_language/vl_sdpa_transformations.cpp (hunk 3 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/vl_sdpa_transformations.cpp]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 4 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp]: done

File: src/cpp/src/visual_language/vl_sdpa_transformations.hpp (hunk 5 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/vl_sdpa_transformations.hpp]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 6 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 7 above)
Comment (rkazants) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: Still not resolved?

File: src/cpp/src/visual_language/qwen2vl/classes.cpp [hunk 9]
```diff
@@ -319,6 +320,29 @@ ov::Tensor get_attention_mask(const std::vector<std::array<size_t, 3>>& reordere
     return attention_mask;
 }
 
+ov::Tensor get_cu_seqlens(const std::vector<std::array<size_t, 3>>& reordered_images_grid_thw) {
```
Comment (rkazants) [inline src/cpp/src/visual_language/qwen2vl/classes.cpp:235]: still not resolved?

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 7 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: No, I don't have it in C++ implementation. The memory of "attention_mask" is nonnegligible in long context.

File: src/cpp/src/visual_language/vl_sdpa_transformations.hpp (hunk 8 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/vl_sdpa_transformations.hpp]: done

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 9 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:235]: Either attention_mask or cu_seqlens, it depends on plugin's capability to support VLSDPA. "cu_seqlens" is preferable to "attention_mask" in terms of memory efficiency. Why do we have to keep a redundant "attention_mask" and then map it to "cu_seqlens" which may negatively impact performance too?

Comment (rkazants) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:235]: My ask is to do this computation in OV runtime, not doing it in C++. I do not propose to replace cu_seqlens with attention_mask. You can replace `cu_seqlens` with `reordered_images_grid_thw` ov::Tensor

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 7 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: Let's compare the following two choices:
1. c++ code of "get_attention_mask(reordered_images_grid_thw)" -> map attention_mask to cu_seqlens and execute it as a part of main graph -> use cu_seqlens in plugin
2. c++ code of "get_cu_seqlens(reordered_images_grid_thw)" -> use cu_seqlens in plugin

Then you see the second choice is more preferable. Do you agree?

Comment (rkazants) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: you don't need to compute attention_mask. You should compute `cu_seqlens` directly from `reordered_images_grid_thw`. 
For this you need to create ov::Model, compile it and execute in OV runtime.
You can also stitch this graph to the main one.

Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: I got it. So it is better to make either get_cu_seqlens or get_attention_mask be part of graph. How about we handle this piece of preprocessing work in another ticket? This PR is mainly for VLSDPA optimizing.

Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:252]: CVS-170675 created to follow up this issue.

**Review by Wovchena — COMMENTED** [review]
Please check accuracy with https://github.com/openvinotoolkit/openvino.genai/tree/master/tools/who_what_benchmark triggering the mew implementation to ensure it doesn't break accuracy. We may end up in a situation like in ticket 170624 otherwise.

I have i9-12900K. Do I have a chance to trigger the new implementation?

File: src/cpp/src/visual_language/qwen2vl/classes.cpp (hunk 9 above)
Comment (ceciliapeng2011) [reply src/cpp/src/visual_language/qwen2vl/classes.cpp:235]: [CVS-170675](https://jira.devtools.intel.com/browse/CVS-170675) created to follow up this issue.

**Comment by ceciliapeng2011** [comment]
> Please check accuracy with https://github.com/openvinotoolkit/openvino.genai/tree/master/tools/who_what_benchmark triggering the mew implementation to ensure it doesn't break accuracy. We may end up in a situation like in ticket 170624 otherwise.
> 
> I have i9-12900K. Do I have a chance to trigger the new implementation?

Will do
